wagtailinventory/views.py

```diff
@@ -1,35 +1,65 @@
-from django.shortcuts import render
-from django.views.generic import View
+from django import forms
 
+from wagtail.admin.filters import WagtailFilterSet
+from wagtail.admin.views.reports import PageReportView
 from wagtail.models import Page
 
-from wagtailinventory.forms import PageBlockQueryFormSet
+import django_filters
 
+from wagtailinventory.checks import dal_select2_check_all
+from wagtailinventory.models import PageBlock
 
-class SearchView(View):
-    template_name = "wagtailinventory/search.html"
 
-    def get(self, request):
-        if len(request.GET) > 1:
-            formset = PageBlockQueryFormSet(request.GET)
-            if not formset.is_valid():
-                return HttpResponseBadRequest("invalid query")
+# If django-autocomplete-light is present, we'll do two things:
+#
+# 1. Set the widget used for the block include/exclude filters to DAL's
+#    multi-select;
+# 2. Create a list view populated with available blocks to autocomplete.
+#
+# If it's not present, we'll fall back on Django's built-in SelectMultiple
+# widget.
+if not dal_select2_check_all():  # pragma: no cover
+    autocomplete = None
+    block_widget = forms.SelectMultiple()
+else:
+    from dal import autocomplete
 
-            queryset = formset.get_query()
-        else:
-            formset = PageBlockQueryFormSet()
-            queryset = Page.objects.all()
+    block_widget = autocomplete.Select2Multiple(url="block-autocomplete")
 
-        queryset = queryset.order_by("title")
+    class BlockAutocompleteView(autocomplete.Select2ListView):
+        def get_list(self):
+            return (
+                PageBlock.objects.distinct()
+                .order_by("block")
+                .values_list("block", flat=True)
+            )
 
-        paginator = Paginator(queryset, per_page=20)
-        pages = paginator.get_page(request.GET.get("p"))
 
-        for page in pages:
-            page.can_choose = True
+class BlockInventoryFilterSet(WagtailFilterSet):
+    include_page_blocks = django_filters.AllValuesMultipleFilter(
+        field_name="page_blocks__block",
+        label="Include Blocks",
+        distinct=True,
+        widget=block_widget,
+    )
+    exclude_page_blocks = django_filters.AllValuesMultipleFilter(
+        field_name="page_blocks__block",
+        label="Exclude Blocks",
+        exclude=True,
+        widget=block_widget,
+    )
 
-        return render(
-            request, self.template_name, {"formset": formset, "pages": pages}
-        )
+    class Meta:
+        model = Page
+        fields = ["include_page_blocks", "exclude_page_blocks"]
+
+
+class BlockInventoryReportView(PageReportView):
+    template_name = "wagtailinventory/inventory_report.html"
+    title = "Block inventory"
+    header_icon = "placeholder"
+    filterset_class = BlockInventoryFilterSet
+
+    def get_queryset(self):
+        self.queryset = Page.objects.all()
+        return super().get_queryset()
```

Comment:
is this logic intentional? Should this just be `return Page.objects.all()` or is there some other reason this sets the queryset and then calls super?

Reply:
Like the template above, there was some iteration here. However, returning `super().get_queryset()` after setting `self.queryset` [is a pattern](https://github.com/wagtail/wagtail/blob/stable/3.0.x/wagtail/admin/views/reports/aging_pages.py#L79-L92) in multiple [Wagtail reports](https://github.com/wagtail/wagtail/blob/stable/3.0.x/wagtail/admin/views/reports/locked_pages.py#L59-L60).

It looks like the `get_queryset()` eventually comes from Django's [`MultipleObjectMixin`](https://github.com/django/django/blob/db7bb3b64e469fbb5c79e7b5b2fcb890434aa60f/django/views/generic/list.py#L22-L47) which does some things on `self.queryset`, including ordering. However, it also looks like it defaults to [`self.model._default_manager.all()`](https://github.com/django/django/blob/db7bb3b64e469fbb5c79e7b5b2fcb890434aa60f/django/views/generic/list.py#L34), so this entire method could probably be removed since it doesn't do anything to the page queryset. I'll try that.

Reply:
Removed in [9c8cbc9](https://github.com/cfpb/wagtail-inventory/pull/68/commits/9c8cbc9f864e31d553c4fa6a677f138912a95ed7).

Reply:
I kept it because it looks like the only place to add our custom ordering is here.